bitcoin/src/Bitcoin/Crypto/Signature.hs

```diff
@@ -14,55 +14,35 @@ module Bitcoin.Crypto.Signature (
     verifyHashSig,
     isCanonicalHalfOrder,
     decodeStrictSig,
-    exportSig,
 ) where
 
-import Bitcoin.Crypto.Hash (Hash256)
+import Bitcoin.Crypto.Hash (Hash256 (getHash256))
 import qualified Bitcoin.Util as U
 import Control.Monad (guard, unless, when)
-import Crypto.Secp256k1 (
-    CompactSig (getCompactSig),
-    Msg,
-    PubKey,
-    SecKey,
-    Sig,
-    exportCompactSig,
-    exportSig,
-    importSig,
-    msg,
-    normalizeSig,
-    signMsg,
-    verifySig,
- )
+import Crypto.Secp256k1 (PubKeyXY, SecKey, Signature, ecdsaNormalizeSignature, ecdsaSign, ecdsaVerify, exportSignatureCompact, exportSignatureDer, importSignatureDer)
 import Data.Binary.Get (Get, getByteString, getWord8, lookAhead)
 import Data.Binary.Put (Put, putByteString)
 import Data.ByteString (ByteString)
 import qualified Data.ByteString as BS
+import Data.ByteString.Short (fromShort)
 import Data.Maybe (fromMaybe, isNothing)
 import Numeric (showHex)
 
 
--- | Convert 256-bit hash into a 'Msg' for signing or verification.
-hashToMsg :: Hash256 -> Msg
-hashToMsg = fromMaybe e . msg . U.encodeS
-  where
-    e = error "Could not convert 32-byte hash to secp256k1 message"
-
-
 -- | Sign a 256-bit hash using secp256k1 elliptic curve.
-signHash :: SecKey -> Hash256 -> Sig
-signHash k = signMsg k . hashToMsg
+signHash :: SecKey -> Hash256 -> Maybe Signature
+signHash k = ecdsaSign k . fromShort . getHash256
```

Comment:
Also I'd like some feedback here around whether the very low probability of signature failure should be modeled as a Maybe or if we should be naughty and just make it a runtime error.

Reply:
Why not both?  I think we definitely want the "main" version to be pure and handle the error case properly.  Because of laziness, impure errors are especially dangerous, IMO.  However, we can also export `signHashUnsafe` (or something) which calls `throw`.

My goto for this used to be `decodeUtf8` and `decodeUtf8'` from `text`, but after a recent change, you have to explicitly say what error treatment you want.

Reply:
Cool. I ask only because I updated the core `libsecp256k1` function to reflect the failure possibility. However, pulling from the C Lib docs:

```C
/** Create an ECDSA signature.
 *
 *  Returns: 1: signature created
 *           0: the nonce generation function failed, or the private key was invalid.
 *  Args:    ctx:    pointer to a context object, initialized for signing (cannot be NULL)
 *  Out:     sig:    pointer to an array where the signature will be placed (cannot be NULL)
 *  In:      msg32:  the 32-byte message hash being signed (cannot be NULL)
 *           seckey: pointer to a 32-byte secret key (cannot be NULL)
 *           noncefp:pointer to a nonce generation function. If NULL, secp256k1_nonce_function_default is used
 *           ndata:  pointer to arbitrary data used by the nonce generation function (can be NULL)
 *
 * The created signature is always in lower-S form. See
 * secp256k1_ecdsa_signature_normalize for more details.
 */
```

The two failure conditions are either:
1. that the private key is invalid which we should be able to exclude using the `importSecKey` function
2. nonce generation failed and we currently leave that up to the C-libsecp256k1 default (which itself currently uses RFC6979), so I'm fairly certain that cannot fail.

So it makes me wonder if I can actually get away with ditching the `Maybe` here.

Reply:
I lean towards ditching the `Maybe` if there isn't a real case of failure. It sounds like this library, if used correctly, would never result in a `Maybe`.

Reply:
I think that's right. I'll go ahead and update this PR to reflect that.

Reply:
OK, agreed.